### ISSUE TITLE
Gateway/API: v2: Discord API v8 preparations

### ISSUE THREAD

**Issue by mavolin** [diamondburned/arikawa#151]
With Discord releasing gateway and API version 8, arikawa should implement the changes as part of arikawa v2.
A list of all changes can be found [here](https://discord.com/developers/docs/change-log#api-and-gateway-v8).

I suggest making a v2 branch that is merged into master once the implementation is complete.

**Comment by diamondburned** [comment]
A [v2 branch](https://github.com/diamondburned/arikawa/tree/v2) has been made.

**Comment by diamondburned** [comment]
I suggest this issue to be kept open for questions and comments regarding Arikawa v2 and Discord v8 API questions.

**Comment by diamondburned** [comment]
### Remove thread-safe guarantees from voice UDP connections

v2 should remove mutexes, channels and other synchronization primitives from package `voice/udp`. The caller should guarantee that only a single goroutine is calling `Write`, and synchronizing calls may be a bit too expensive or overkill.

**Comment by karitham** [comment]
### Make automatic reconnexion optional

Automatic reconnexion is nice, but sometime doesn't work and the bot never ends up reconnecting, or reconnects after a while

Having a an option to stop the bot from automatically reconnecting would be a great idea as it would allow for us to handle the reconnexion ourselves, or just make the bot restart

**Comment by mavolin** [comment]
Instead of just allowing to fully disable auto reconnects, why not allow to set the number of maximum retries? `> 1` could be unlimited, `>= 0` is interpreted as is.

**Comment by diamondburned** [comment]
I think a deadline is more reasonable than a retry time, but it shouldn't be a huge difference, as connections themselves already have deadlines (I think?).

**Comment by mavolin** [comment]
Seeing as Gateway Intents are now required, I would propose to remove all constructor funcs, etc, that don't take in `Intents`, but are required per discord from now on.

**Comment by diamondburned** [comment]
This library isn't strictly for bots only, so I don't think that's a good idea.

**Comment by mavolin** [comment]
Forgot about that

**Comment by diamondburned** [comment]
API v8 is so far completely implemented except for slash commands, which has its own tracker #176. Closing this issue.
